src/main/resources/session-template-data/template.ftl

```diff
@@ -1,121 +1,161 @@
+
+        -- insert data into real session location group table from temporary one.
+        INSERT INTO session_location_group(id,reference,prison_id,name)
+                    SELECT id,CONCAT('-',REGEXP_REPLACE(to_hex((ROW_NUMBER () OVER (ORDER BY key))+2951597050), '(.{3})(?!$)', '\1~','g')) as reference,prison_id,name FROM tmp_session_location_group order by id;
+
+        -- Sequence updated manually as id's were inserted from temp table
+        ALTER SEQUENCE session_location_group_id_seq RESTART WITH  ${session_location_group_id_index};
+
+
+        -- Create permitted session location data
+        CREATE TABLE tmp_permitted_session_location (
+            id                serial        NOT NULL PRIMARY KEY,
+            group_key        VARCHAR(100)  NOT NULL,
+            group_id          int,
+            level_one_code    VARCHAR(10) NOT NULL,
+            level_two_code    VARCHAR(10),
+            level_three_code  VARCHAR(10),
+            level_four_code   VARCHAR(10)
+        );
+
+        INSERT INTO tmp_permitted_session_location (group_key,level_one_code,level_two_code,level_three_code,level_four_code)
+        VALUES
+        <#list locations as l>
+            ('${l.groupKey}','${l.levelOne}',<#if l.levelTwo??>'${l.levelTwo}'<#else>NULL</#if>,<#if l.levelThree??>'${l.levelThree}'<#else>NULL</#if>, <#if l.levelFour??>'${l.levelFour}'<#else>NULL</#if>)<#if l_has_next>,</#if>
+        </#list>;
+
+        -- update tmp location table with correct group_id.
+        UPDATE tmp_permitted_session_location SET group_id = tmp_session_location_group.id FROM tmp_session_location_group WHERE tmp_permitted_session_location.group_key = tmp_session_location_group.key;
+
+
+        -- insert data into real permitted session location table from temporary one.
+        INSERT INTO permitted_session_location(id,group_id,level_one_code,level_two_code,level_three_code,level_four_code)
+            SELECT id,group_id,level_one_code,level_two_code,level_three_code,level_four_code FROM tmp_permitted_session_location order by id;
+
+        ALTER SEQUENCE permitted_session_location_id_seq RESTART WITH ${permitted_session_location_index};
+
+
+        -- Create link table data
+        INSERT INTO session_to_location_group(session_template_id, group_id)
+        SELECT st.id, g.id FROM tmp_session_template st
+                                           JOIN tmp_session_location_group g ON POSITION(g.key  IN st.locationKeys)<>0 ORDER BY st.id,g.id;
+
+        -- Location group names are only descriptions they need to be updated when the group locations change
+
+        -- Prisoner category
+
+        -- Create temporary group table
+            CREATE TABLE tmp_session_prisoner_category (
+                id                  serial          NOT NULL PRIMARY KEY,
+                code                varchar(100)    NOT NULL
+            );
+
+        <#if categories?has_content>
```

Comment:
just for clarity - will this create duplicate codes? Should code be unique on the table so that its constant across prisons?

Reply:
No duplicates the code puts the categories in a set in the SQL generator!